File: SignalServiceKit/src/Messages/OWSOutgoingReceiptManager.m
//
// Copyright 2018 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

#import "OWSOutgoingReceiptManager.h"
#import "AppReadiness.h"
#import "FunctionalUtil.h"
#import "MessageSender.h"
#import "OWSError.h"
#import "OWSReceiptsForSenderMessage.h"
#import "TSContactThread.h"
#import "TSYapDatabaseObject.h"
#import <SignalServiceKit/SignalServiceKit-Swift.h>

NS_ASSUME_NONNULL_BEGIN

NSString *NSStringForOWSReceiptType(OWSReceiptType receiptType)
{
    switch (receiptType) {
        case OWSReceiptType_Delivery:
            return @"Delivery";
        case OWSReceiptType_Read:
            return @"Read";
        case OWSReceiptType_Viewed:
            return @"Viewed";
        default:
            return @"Unknown";
    }
}

#pragma mark -

@interface OWSOutgoingReceiptManager ()

// This property should only be accessed on the serialQueue.
@property (nonatomic) BOOL isProcessing;

@end

#pragma mark -

@implementation OWSOutgoingReceiptManager

+ (SDSKeyValueStore *)deliveryReceiptStore
{
    return [[SDSKeyValueStore alloc] initWithCollection:@"kOutgoingDeliveryReceiptManagerCollection"];
}

+ (SDSKeyValueStore *)readReceiptStore
{
    return [[SDSKeyValueStore alloc] initWithCollection:@"kOutgoingReadReceiptManagerCollection"];
}

+ (SDSKeyValueStore *)viewedReceiptStore
{
    return [[SDSKeyValueStore alloc] initWithCollection:@"kOutgoingViewedReceiptManagerCollection"];
}

#pragma mark -

- (instancetype)init
{
    self = [super init];

    if (!self) {
        return self;
    }

    OWSSingletonAssert();

    _pendingTasks = [[PendingTasks alloc] initWithLabel:@"Receipt Sends"];

    // We skip any sends to untrusted identities since we know they'll fail anyway. If an identity state changes
    // we should recheck our pendingReceipts to re-attempt a send to formerly untrusted recipients.
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(process)
                                                 name:OWSIdentityManagerObjCBridge.identityStateDidChangeNotification
                                               object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(process)
                                                 name:SSKReachability.owsReachabilityDidChange
                                               object:nil];

    // Start processing.
    AppReadinessRunNowOrWhenAppDidBecomeReadyAsync(^{ [self process]; });

    return self;
}

- (dispatch_queue_t)serialQueue
{
    static dispatch_queue_t _serialQueue;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _serialQueue
            = dispatch_queue_create("org.signal.outgoing-receipts", DISPATCH_QUEUE_SERIAL_WITH_AUTORELEASE_POOL);
    });

    return _serialQueue;
}

// Schedules a processing pass, unless one is already scheduled.
- (void)process {
    [self processWithCompletion:nil];
}

// Schedules a processing pass, unless one is already scheduled.
- (void)processWithCompletion:(void (^_Nullable)(void))completion
{
    if (!AppReadiness.isAppReady && !CurrentAppContext().isRunningTests) {
        OWSFailDebug(@"Outgoing receipts require app is ready");
        if (completion) {
            completion();
        }
        return;
    }

    dispatch_async(self.serialQueue, ^{
        if (self.isProcessing) {
            if (completion) {
                completion();
            }
            [self logMemoryUsage];
            return;
        }

        OWSLogVerbose(@"Processing outbound receipts.");

        self.isProcessing = YES;

        if (!self.reachabilityManager.isReachable) {
            // No network availability; abort.
            self.isProcessing = NO;
            if (completion) {
                completion();
            }
            [self logMemoryUsage];
            return;
        }

        NSMutableArray<AnyPromise *> *sendPromises = [NSMutableArray array];
        [sendPromises addObjectsFromArray:[self sendReceiptsForReceiptType:OWSReceiptType_Delivery]];
        [sendPromises addObjectsFromArray:[self sendReceiptsForReceiptType:OWSReceiptType_Read]];
        [sendPromises addObjectsFromArray:[self sendReceiptsForReceiptType:OWSReceiptType_Viewed]];

        if (sendPromises.count < 1) {
            // No work to do; abort.
            self.isProcessing = NO;
            if (completion) {
                completion();
            }
            [self logMemoryUsage];
            return;
        }

        AnyPromise *completionPromise = [AnyPromise whenResolved:sendPromises];
        completionPromise.ensure(^() {
            if (completion) {
                completion();
            }
            // Wait N seconds before conducting another pass.
            // This allows time for a batch to accumulate.
            //
            // We want a value high enough to allow us to effectively de-duplicate
            // receipts without being so high that we incur so much latency that
            // the user notices.
            const CGFloat kProcessingFrequencySeconds = 3.f;
            dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(kProcessingFrequencySeconds * NSEC_PER_SEC)),
                self.serialQueue,
                ^{
                    self.isProcessing = NO;

                    [self process];
                });
        });
    });
}

- (void)logMemoryUsage {
    if (SSKDebugFlags.internalLogging) {
        OWSLogInfo(@"memoryUsage: %@", LocalDevice.memoryUsageString);
    }
}

- (NSArray<AnyPromise *> *)sendReceiptsForReceiptType:(OWSReceiptType)receiptType {
    __block NSDictionary<SignalServiceAddress *, MessageReceiptSet *> *queuedReceiptMap;
    [self.databaseStorage
        readWithBlock:^(SDSAnyReadTransaction *transaction) {
            NSMutableDictionary *receiptSetsToSend = [[self fetchAllReceiptSetsWithType:receiptType
                                                                            transaction:transaction] mutableCopy];
            NSArray *excludedAddresses = [receiptSetsToSend.allKeys filter:^BOOL(SignalServiceAddress *address) {
                return [self.blockingManager isAddressBlocked:address transaction:transaction] ||
                    [RecipientHidingManagerObjcBridge isHiddenAddress:address tx:transaction];
            }];

            for (SignalServiceAddress *address in excludedAddresses) {
                OWSLogWarn(@"Skipping send for excluded address: %@", address);
                // If an address is excluded, we don't bother sending a receipt.
                // We remove it from our fetched list, and dequeue it from our pending receipt set
                MessageReceiptSet *receiptSet = receiptSetsToSend[address];
                [self dequeueReceiptsFor:address receiptSet:receiptSet receiptType:receiptType];
                [receiptSetsToSend removeObjectForKey:address];
            }
            queuedReceiptMap = [receiptSetsToSend copy];
        }
                 file:__FILE__
             function:__FUNCTION__
                 line:__LINE__];

    NSMutableArray<AnyPromise *> *sendPromises = [NSMutableArray array];

    for (SignalServiceAddress *address in queuedReceiptMap) {
        if (!address.isValid) {
            OWSFailDebug(@"Unexpected identifier.");
            continue;
        }

        MessageReceiptSet *receiptSet = queuedReceiptMap[address];
        if (receiptSet.timestamps.count < 1) {
            OWSFailDebug(@"Missing timestamps.");
            continue;
        }

        if ([OWSIdentityManagerObjCBridge untrustedIdentityForSendingToAddress:address]) {
            OWSLogWarn(@"%@ is untrusted. Deferring sending of receipts.", address);
            continue;
        }

        TSThread *thread = [TSContactThread getOrCreateThreadWithContactAddress:address];

        __block AnyPromise *sendPromise;
        DatabaseStorageWrite(self.databaseStorage, ^(SDSAnyWriteTransaction *transaction) {
            OWSReceiptsForSenderMessage *message;
            NSString *receiptName = NSStringForOWSReceiptType(receiptType);
            switch (receiptType) {
                case OWSReceiptType_Delivery:
                    message = [OWSReceiptsForSenderMessage deliveryReceiptsForSenderMessageWithThread:thread
                                                                                           receiptSet:receiptSet
                                                                                          transaction:transaction];
                    break;
                case OWSReceiptType_Read:
                    message = [OWSReceiptsForSenderMessage readReceiptsForSenderMessageWithThread:thread
                                                                                       receiptSet:receiptSet
                                                                                      transaction:transaction];
                    break;
                case OWSReceiptType_Viewed:
                    message = [OWSReceiptsForSenderMessage viewedReceiptsForSenderMessageWithThread:thread
                                                                                         receiptSet:receiptSet
                                                                                        transaction:transaction];
                    break;
            }

            sendPromise =
                [self.sskJobQueues.messageSenderJobQueue addPromiseWithMessage:message.asPreparer
                                                     removeMessageAfterSending:NO
                                                 limitToCurrentProcessLifetime:YES
                                                                isHighPriority:NO
                                                                   transaction:transaction]
                    .doneInBackground(^(id value) {
                        OWSLogInfo(@"Successfully sent %lu %@ receipts to sender.",
                            (unsigned long)receiptSet.timestamps.count,
                            receiptName);

                        // DURABLE CLEANUP - we could replace the custom durability logic in this class
                        // with a durable JobQueue.
                        [self dequeueReceiptsFor:address receiptSet:receiptSet receiptType:receiptType];
                    })
                    .catchInBackground(^(NSError *error) {
                        OWSLogError(@"Failed to send %@ receipts to sender with error: %@", receiptName, error);

                        if ([MessageSenderNoSuchSignalRecipientError isNoSuchSignalRecipientError:error]) {
                            [self dequeueReceiptsFor:address receiptSet:receiptSet receiptType:receiptType];
                        }
                    });
        });

        [sendPromises addObject:sendPromise];
    }

    return [sendPromises copy];
}

- (SDSKeyValueStore *)storeForReceiptType:(OWSReceiptType)receiptType
{
    switch (receiptType) {
        case OWSReceiptType_Delivery:
            return OWSOutgoingReceiptManager.deliveryReceiptStore;
        case OWSReceiptType_Read:
            return OWSOutgoingReceiptManager.readReceiptStore;
        case OWSReceiptType_Viewed:
            return OWSOutgoingReceiptManager.viewedReceiptStore;
    }
}

@end

NS_ASSUME_NONNULL_END

File: SignalServiceKit/src/Messages/OWSOutgoingReceiptManager.swift
//
// Copyright 2021 Signal Messenger, LLC
// SPDX-License-Identifier: AGPL-3.0-only
//

import Foundation
import LibSignalClient
import SignalCoreKit

@objc
class MessageReceiptSet: NSObject, Codable {
    @objc
    public private(set) var timestamps: Set<UInt64> = Set()
    @objc
    public private(set) var uniqueIds: Set<String> = Set()

    @objc
    func insert(timestamp: UInt64, messageUniqueId: String? = nil) {
        timestamps.insert(timestamp)
        if let uniqueId = messageUniqueId {
            uniqueIds.insert(uniqueId)
        }
    }

    @objc
    func union(_ other: MessageReceiptSet) {
        timestamps.formUnion(other.timestamps)
        uniqueIds.formUnion(other.uniqueIds)
    }

    @objc
    func subtract(_ other: MessageReceiptSet) {
        timestamps.subtract(other.timestamps)
        uniqueIds.subtract(other.uniqueIds)
    }

    fileprivate func union(timestampSet: Set<UInt64>) {
        timestamps.formUnion(timestampSet)
    }
}

extension OWSOutgoingReceiptManager {
    @objc
    func enqueueDeliveryReceipt(
        for decryptedEnvelope: DecryptedIncomingEnvelope,
        messageUniqueId: String?,
        tx: SDSAnyWriteTransaction
    ) {
        enqueueReceipt(
            for: SignalServiceAddress(decryptedEnvelope.sourceAci),
            timestamp: decryptedEnvelope.timestamp,
            messageUniqueId: messageUniqueId,
            receiptType: .delivery,
            tx: tx
        )
    }

    @objc
    public func enqueueReadReceipt(
        for address: SignalServiceAddress,
        timestamp: UInt64,
        messageUniqueId: String?,
        tx: SDSAnyWriteTransaction
    ) {
        enqueueReceipt(
            for: address,
            timestamp: timestamp,
            messageUniqueId: messageUniqueId,
            receiptType: .read,
            tx: tx
        )
    }

    @objc
    public func enqueueViewedReceipt(
        for address: SignalServiceAddress,
        timestamp: UInt64,
        messageUniqueId: String?,
        tx: SDSAnyWriteTransaction
    ) {
        enqueueReceipt(
            for: address,
            timestamp: timestamp,
            messageUniqueId: messageUniqueId,
            receiptType: .viewed,
            tx: tx
        )
    }

    private func enqueueReceipt(
        for address: SignalServiceAddress,
        timestamp: UInt64,
        messageUniqueId: String?,
        receiptType: OWSReceiptType,
        tx: SDSAnyWriteTransaction
    ) {
        owsAssertDebug(address.isValid)
        guard timestamp >= 1 else {
            owsFailDebug("Invalid timestamp.")
            return
        }
        let isHiddenRecipient = FeatureFlags.recipientHiding
            && DependenciesBridge.shared.recipientHidingManager.isHiddenAddress(address, tx: tx.asV2Read)
        guard !isHiddenRecipient else {
            // Don't send receipts to hidden recipients.
            return
        }
        let pendingTask = pendingTasks.buildPendingTask(label: "Receipt Send: \(NSStringForOWSReceiptType(receiptType))")
        let persistedSet = fetchAndMergeReceiptSet(type: receiptType, address: address, transaction: tx)
        persistedSet.insert(timestamp: timestamp, messageUniqueId: messageUniqueId)
        storeReceiptSet(persistedSet, type: receiptType, address: address, transaction: tx)
        tx.addAsyncCompletionOffMain { self.process(completion: { pendingTask.complete() }) }
    }

    @objc
    func dequeueReceipts(for address: SignalServiceAddress, receiptSet: MessageReceiptSet, receiptType: OWSReceiptType) {
        owsAssertDebug(address.isValid)
        databaseStorage.asyncWrite { tx in
            let persistedSet = self.fetchAndMergeReceiptSet(type: receiptType, address: address, transaction: tx)
            persistedSet.subtract(receiptSet)
            self.storeReceiptSet(persistedSet, type: receiptType, address: address, transaction: tx)
        }
    }

    @objc
    func fetchAllReceiptSets(type: OWSReceiptType, transaction tx: SDSAnyReadTransaction) -> [SignalServiceAddress: MessageReceiptSet] {
        return Dictionary(
            uniqueKeysWithValues: (
                Set(store(for: type).allKeys(transaction: tx).compactMap { SignalServiceAddress(identifier: $0) })
                    .map { ($0, fetchReceiptSet(type: type, preferredKey: $0.aciUppercaseString, secondaryKey: $0.phoneNumber, tx: tx).receiptSet) }
            )
        )
    }

    func fetchAndMergeReceiptSet(type: OWSReceiptType, address: SignalServiceAddress, transaction tx: SDSAnyWriteTransaction) -> MessageReceiptSet {
        return fetchAndMerge(type: type, preferredKey: address.aciUppercaseString, secondaryKey: address.phoneNumber, tx: tx)
    }

    private func fetchReceiptSet(
        type: OWSReceiptType,
        preferredKey: String?,
        secondaryKey: String?,
        tx: SDSAnyReadTransaction
    ) -> (receiptSet: MessageReceiptSet, hasSecondaryValue: Bool) {
        let store = store(for: type)
        let result = MessageReceiptSet()
        if let preferredKey, store.hasValue(forKey: preferredKey, transaction: tx) {
            if let receiptSet: MessageReceiptSet = try? store.getCodableValue(forKey: preferredKey, transaction: tx) {
                result.union(receiptSet)
            } else if let numberSet = store.getObject(forKey: preferredKey, transaction: tx) as? Set<UInt64> {
                result.union(timestampSet: numberSet)
            }
        }
        var hasSecondaryValue = false
        if let secondaryKey, store.hasValue(forKey: secondaryKey, transaction: tx) {
            if let receiptSet: MessageReceiptSet = try? store.getCodableValue(forKey: secondaryKey, transaction: tx) {
                result.union(receiptSet)
            } else if let numberSet = store.getObject(forKey: secondaryKey, transaction: tx) as? Set<UInt64> {
                result.union(timestampSet: numberSet)
            }
            hasSecondaryValue = true
        }
        return (result, hasSecondaryValue)
    }

    private func fetchAndMerge(
        type: OWSReceiptType,
        preferredKey: String?,
        secondaryKey: String?,
        tx: SDSAnyWriteTransaction
    ) -> MessageReceiptSet {
        let (result, hasSecondaryValue) = fetchReceiptSet(
            type: type,
            preferredKey: preferredKey,
            secondaryKey: secondaryKey,
            tx: tx
        )
        if let preferredKey, let secondaryKey, hasSecondaryValue {
            store(for: type).removeValue(forKey: secondaryKey, transaction: tx)
            _storeReceiptSet(result, type: type, key: preferredKey, tx: tx)
        }
        return result
    }

    @objc
    func storeReceiptSet(_ receiptSet: MessageReceiptSet, type: OWSReceiptType, address: SignalServiceAddress, transaction: SDSAnyWriteTransaction) {
        guard let key = address.aciUppercaseString ?? address.phoneNumber else {
            return owsFailDebug("Invalid address")
        }
        _storeReceiptSet(receiptSet, type: type, key: key, tx: transaction)
    }

    private func _storeReceiptSet(_ receiptSet: MessageReceiptSet, type: OWSReceiptType, key: String, tx: SDSAnyWriteTransaction) {
        let store = store(for: type)
        if receiptSet.timestamps.count > 0 {
            do {
                try store.setCodable(receiptSet, key: key, transaction: tx)
            } catch {
                owsFailDebug("\(error)")
            }
        } else {
            store.removeValue(forKey: key, transaction: tx)
        }
    }

    public func pendingSendsPromise() -> Promise<Void> {
        // This promise blocks on all operations already in the queue,
        // but will not block on new operations added after this promise
        // is created. That's intentional to ensure that NotificationService
        // instances complete in a timely way.
        pendingTasks.pendingTasksPromise()
    }
}

// MARK: -

fileprivate extension SignalServiceAddress {
    convenience init?(identifier: String) {
        if let aci = Aci.parseFrom(aciString: identifier) {
            self.init(aci)
        } else if identifier.isStructurallyValidE164 {
            self.init(phoneNumber: identifier)
        } else {
            return nil
        }
    }
}
